### ISSUE TITLE
etcd plugin not resolving CNAME entries

### ISSUE THREAD

**Issue by lholota** [miekg/coredns#7028]
Hi all,
I have a stack set up with external-dns -> etcd -> coredns via etcd plugin. A records resolve correctly, but all CNAME records only return NOERROR and no value even though the CNAME's target is an A record existing in the same etcd as shown below:

When I checked the etcd plugin documentation, it doesn't mention CNAME records except for records pointing to external names. The Corefile and etcd state are below. Are CNAMEs not supported or am I doing something wrong? The whole external-dns is built around CNAMEs so I can't really imagine why its developers would support coredns via etcd/skydns if one of the core usecases wasn't supported...

```
.:53 {
    errors
    health {
        lameduck 5s
    }
    ready
    prometheus 0.0.0.0:9153
    log
    etcd example.com {
        stubzones
        path /skydns
        endpoint http://external-dns-etcd.external-dns-system.svc.cluster.local:2379
        credentials root <password redacted>
        upstream
    }
    forward . /etc/resolv.conf
    cache 30
    loop
    reload
    loadbalance
}
```

```
/skydns/com/example/a-example2/00e24061
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=crd/external-dns-
system/example\"","targetstrip":1}
/skydns/com/example/a-example2/2c23e944
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=crd/external-dns-system/example\"","targetstrip":1}
/skydns/com/example/a-traefik/1d8b7880
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=service/traefik-system/traefik\"","targetstrip":1}
/skydns/com/example/a-traefik/664f46bc
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=service/traefik-system/traefik\"","targetstrip":1}
/skydns/com/example/cname-example1/2f6d448d
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=crd/external-dns-system/example\"","targetstrip":1}
/skydns/com/example/cname-oauth/0108beb1
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=ingressroute/traefik-system/traefik-oauth2-proxy\"","targetstrip":1}
/skydns/com/example/cname-oauth/1c1d4220
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=ingressroute/traefik-system/traefik-oauth2-proxy\"","targetstrip":1}
/skydns/com/example/cname-oauth/583611e5
{"text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=ingressroute/traefik-system/traefik-oauth2-proxy\"","targetstrip":1}
/skydns/com/example/example1/270896aa
{"host":"example2.example.com","text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=crd/external-dns-system/example\"","ttl":1,"targetstrip":1}
/skydns/com/example/example2/1ba35cde
{"host":"1.1.1.1","text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=crd/external-dns-system/example\"","ttl":1,"targetstrip":1}
/skydns/com/example/oauth/6dec8b0b
{"host":"traefik.example.com","text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=ingressroute/traefik-system/traefik-oauth2-proxy\"","ttl":1,"targetstrip":1}
/skydns/com/example/traefik/417b21b7
{"host":"10.1.2.128","text":"\"heritage=external-dns,external-dns/owner=default,external-dns/resource=service/traefik-system/traefik\"","targetstrip":1}
```

**Comment by akhilsingh-git** [comment]
The etcd plugin for CoreDNS uses a specific data model to store DNS records. Each record in etcd is a JSON object with keys like host, text, ttl, and targetstrip. Here’s how it interprets these fields:

A Record: If the host field is an IP address, CoreDNS returns an A record.
CNAME Record: If the host field is a domain name rather than an IP, CoreDNS returns a CNAME record.
Text (TXT) Record: If the text field is present, CoreDNS may return TXT records depending on configuration.
SRV or other Records: If no host is given but additional fields are set, CoreDNS may return SRV or other record types accordingly.
In your case, you have entries like:

`{"host":"example2.example.com","text":"\"heritage=external-dns...\"","ttl":1,"targetstrip":1}`

For a CNAME, host is correct as it’s a domain. However, the targetstrip field could be problematic. targetstrip is meant to remove a certain number of labels from the left side of the host before returning it. If this strips too much, it might leave an incomplete domain name, resulting in CoreDNS not returning the expected CNAME response or the subsequent A record resolution.

**Comment by lholota** [comment]
@akhilsingh-git Thank you for pointing me in the right direction. It wasn't the `targetstrip` property - it was caused by the `text` property. I do not understand why it causes this behavior since I was able to resolve the A record even with the property. But when I tried to resolve the CNAME if either the CNAME record itself or the A record it was pointing to contained the text property it failed with NOERROR as described above.

I have filed a [bug in external-dns repo](https://github.com/kubernetes-sigs/external-dns/issues/4953) to check whether it's a doc omission or a missed test case. By default external-dns uses the text property to store record metadata which can be disabled but it's not mentioned in the docs for the coredns provider.

Based on what I've found at the [skydns repo](https://github.com/skynetservices/skydns), the text property should be used for handling TXT queries. I can't tell from the spec whether it's valid to have a host and text property in the same etcd entry. If it is valid, this could be a bug in the etcd coredns plugin. Either way, thanks a lot! It helped me find the solution after banging my head against the wall for a day and a half 😉
